packages/loader/container-loader/src/container.ts

```diff
@@ -842,6 +842,16 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         try {
             // Ensure that we raise all key events even if one of these throws
             try {
+                // Raise event first, to ensure we capture _lifecycleState before transition.
+                // This gives us a chance to know what errors happened on open vs. on fully loaded container.
+                this.mc.logger.sendTelemetryEvent(
+                    {
+                        eventName: "ContainerDispose",
+                        category: error === undefined ? "generic" : "error",
+                    },
+                    error,
+                );
+
                 // ! Progressing from "closed" to "disposing" is not allowed
                 if (this._lifecycleState !== "closed") {
```

Comment:
@kian-thompson why is this necessary? Why not always transition to disposing?  They're equivalence-classed in this.closed right?

Reply:
This is a tricky nuance because the "closed" and "disposed" states are checked for the "Container.closed" boolean. In this new case, would we consider "disposing" to mean the Container is closed? My thinking was no, but we can't switch the state to non-closed as it may break a check somewhere.